File: Engine/Sources/CardCopilotEngine/Engine/CapMath.swift
public enum CapMath {
    /// Splits a purchase into the portion still earning the accelerated rate and the post-cap portion.
    public static func split(amount: Double, capLimit: Double, usage: Double)
        -> (inCap: Double, overCap: Double) {
        let room = max(0, capLimit - usage)
        let inCap = min(amount, room)
        return (inCap, amount - inCap)
    }

    /// Splits a purchase against multiple racing caps. The bottleneck cap (the one with the least room)
    /// constrains the in-cap amount.
    public static func splitMulti(amount: Double, caps: [(limit: Double, usage: Double)])
        -> (inCap: Double, overCap: Double) {
        let minRoom = caps.map { max(0, $0.limit - $0.usage) }.min() ?? Double.infinity
        let inCap = min(amount, minRoom)
        return (inCap, amount - inCap)
    }
}

File: android/core/engine/src/main/kotlin/com/cardcopilot/engine/engine/CapMath.kt
package com.cardcopilot.engine.engine

data class CapSplit(val inCap: Double, val overCap: Double)
data class CapInput(val limit: Double, val usage: Double)

object CapMath {
    fun split(amount: Double, capLimit: Double, usage: Double): CapSplit {
        val room = maxOf(0.0, capLimit - usage)
        val inCap = minOf(amount, room)
        return CapSplit(inCap, amount - inCap)
    }

    /** The cap with the least remaining room constrains the accelerated portion. */
    fun splitMulti(amount: Double, caps: List<CapInput>): CapSplit {
        val minRoom = caps.minOfOrNull { maxOf(0.0, it.limit - it.usage) }
            ?: Double.POSITIVE_INFINITY
        val inCap = minOf(amount, minRoom)
        return CapSplit(inCap, amount - inCap)
    }
}
